test/e2e/nodeportlocal_test.go

```diff
@@ -0,0 +1,589 @@
+		data["antrea-agent.conf"] = antreaAgentConf
+	}, false, true); err != nil {
+		t.Fatalf("Failed to update NodePortLocal port range: %v", err)
+	}
+}
+
+func validatePortInRange(t *testing.T, nplAnnotations []k8s.NPLAnnotation, start, end int) {
+	for i := range nplAnnotations {
+		if nplAnnotations[i].NodePort > end || nplAnnotations[i].NodePort < start {
+			t.Fatalf("Node port %d not in range: %d - %d", nplAnnotations[i].NodePort, start, end)
+		}
+	}
+}
+
+// TestNPLAddPod tests NodePortLocal functionalities after adding a pod.
+// - Enable NodePortLocal if not already enabled.
+// - Create an nginx Pod and a Service.
+// - Verify that the required NodePortLocal annoation is added in the Pod.
+// - Make sure IPTABLES rule are correctly added in the Node from Antrea Agent Pod.
+// - Create a client Pod and test traffic through netcat.
+// - Delete the nginx Pod and verify that the IPTABLES rules are deleted.
+func TestNPLAddPod(t *testing.T) {
+	skipIfNotIPv4Cluster(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	enableNPLInConfigmap(t, data)
+	r := require.New(t)
+	k8sUtils, err = NewKubernetesUtils(data)
+	r.Nil(err, "Failed to initialize Kubernetes Utils: %v", err)
+
+	node := nodeName(0)
+	testPodName := randName("test-pod-")
+	err = data.createNginxPod(testPodName, node)
+	r.Nil(err, "Error when creating test Pod: %v", err)
+
+	annotation := make(map[string]string)
+	annotation[nplSvcAnnotation] = "true"
+	ipFamily := corev1.IPv4Protocol
+	data.createNginxClusterIPServiceWithAnnotation(false, &ipFamily, annotation)
+
+	nplAnnotationString, testPodIP := getNPLAnnotation(t, data, r, testPodName)
+	var nplAnnotation []k8s.NPLAnnotation
+	json.Unmarshal([]byte(nplAnnotationString), &nplAnnotation)
+
+	clientName := randName("test-client-")
+	if err := data.createBusyboxPodOnNode(clientName, node); err != nil {
+		t.Fatalf("Error creating Pod %s: %v", clientName, err)
+	}
+
+	_, err = data.podWaitForIPs(defaultTimeout, clientName, testNamespace)
+	if err != nil {
+		t.Fatalf("Error waiting for IP for Pod %s: %v", clientName, err)
+	}
+
+	antreaPod, err := data.getAntreaPodOnNode(node)
+	r.Nil(err, "Error when getting Antrea Agent Pod on Node '%s'", node)
+
+	checkForNPLRuleInIPTABLES(t, data, r, nplAnnotation, antreaPod, testPodIP.ipv4.String(), true)
+	validatePortInRange(t, nplAnnotation, defaultStartPort, defaultEndPort)
+	checkTrafficForNPL(t, data, r, nplAnnotation, clientName)
+
+	data.deletePod(testPodName)
+	checkForNPLRuleInIPTABLES(t, data, r, nplAnnotation, antreaPod, testPodIP.ipv4.String(), false)
+}
+
+// TestNPLMultiplePods tests NodePortLocal functionalities after adding multiple Pods.
+func TestNPLMultiplePods(t *testing.T) {
+	skipIfNotIPv4Cluster(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	enableNPLInConfigmap(t, data)
```

Comment:
I would recommend grouping tests when possible (with commong setup and common teardown), and using subtests: https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks. This will save time.

Reply:
Thanks. I'll check it out.

Reply:
Yeah ! I missed it. I'll remove it.